### PR TITLE
Quick Start for Existing users: Update Stats tour

### PR DESCRIPTION
**Pull request by momo-ozawa**
Part of #18387 

## Description

- Updated the Stats task
    - Updated to point to quick action in dashboard or item in site menu, depending on where the user is
    - Removed the step that points from dashboard to site menu 
    - Added autoscroll on the dashboard screen
    - Added a task complete notice
    
 Site menu | Dashboard | Task complete notice
-- | -- | --
 |  | 

    
### Initial screen: site menu

1. Set the initial screen to the site menu
2. Enable quick start
3. Start the on the Stats tour in the **Grow your audience** checklist
4. ✅ The screen should autoscroll / show a blue waypoint on the Stats item in the site menu
5. Switch to the dashboard tab
6. ✅ The Stats button should not display a waypoint
7. Switch back to the site menu tab
8. Tap on the Stats item
9. ✅ A task complete notice should be shown
10. Go back to My Site
11. ✅ The task complete notice should be dismissed

https://user-images.githubusercontent.com/6711616/164453739-ef6534e6-3b20-4dba-849c-e4c35746cd4c.mp4

### Initial screen: dashboard

1. Set the initial screen to the dashboard
2. Enable quick start
3. Start the on the Stats tour in the **Grow your audience** checklist
4. ✅ The screen should autoscroll / show a waypoint on the Stats button in the Quick Actions section
5. Switch to the site menu tab
6. ✅ The Stats item should not display a waypoint
7. Switch back to the dashboard tab
8. Tap on the Stats button
9. ✅ A task complete notice should be shown
10. Go back to My Site
11. ✅ The task complete notice should be dismissed

https://user-images.githubusercontent.com/6711616/164738234-9fed17b8-16b7-478d-919d-38a9e4af7188.mp4

### Smoke testing on iPhone and iPad

1. Enable quick start
2. Go through all the tours
3. ✅ Make sure all the tours work as expected and the waypoints are being displayed on the correct views

## Regression Notes
1. Potential unintended areas of impact
- Quick Start on iPad

4. What I did to test those areas of impact (or what existing automated tests I relied on)
- Tested the above steps manually

5. What automated tests I added (or what prevented me from doing so)
- n/a

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WordPress/Classes/Extensions/UIScrollView+Helpers.swift [hunk 1]
```diff
@@ -44,6 +46,15 @@ extension UIScrollView {
         }
     }
 
+    // MARK: - Horizontal scrollview
+
+    func scrollToEnd(animated: Bool) {
```
Comment (hassaanelgarem) [inline WordPress/Classes/Extensions/UIScrollView+Helpers.swift:84]: [?] Do we need to call `layoutIfNeeded` at the end of this func same as in `scrollToView`?

File: WordPress/Classes/ViewRelated/Blog/Blog Dashboard/BlogDashboardViewController.swift [hunk 2]
```diff
@@ -147,7 +152,22 @@ final class BlogDashboardViewController: UIViewController {
...
+
+            if let info = notification.userInfo,
+               let element = info[QuickStartTourGuide.notificationElementKey] as? QuickStartTourElement {
+
+                switch element {
+                case .setupQuickStart, .removeQuickStart:
+                    self?.loadCardsFromCache()
+                case .stats:
+                    self?.mySiteScrollView?.scrollToTop(animated: true)
+                    self?.mySiteViewController?.additionalSafeAreaInsets = UIEdgeInsets(top: 0, left: 0, bottom: Constants.bottomPaddingForQuickStartNotices, right: 0)
```
Comment (hassaanelgarem) [inline WordPress/Classes/ViewRelated/Blog/Blog Dashboard/BlogDashboardViewController.swift]: [?] Why is this line needed? When I remove it, it seems like nothing differs 🤔

File: WordPress/Classes/ViewRelated/Blog/Blog Dashboard/Cards/Quick Actions/DashboardQuickActionsCardCell.swift [hunk 3]
```diff
@@ -51,11 +51,24 @@ final class DashboardQuickActionsCardCell: UICollectionViewCell, Reusable {
...
+        stopObservingQuickStart()
+    }
+
+    func configure(blog: Blog, viewController: BlogDashboardViewController?, apiResponse: BlogDashboardRemoteEntity?) {
+        guard let viewController = viewController else {
+            return
+        }
+
+        configureQuickActionButtons(for: blog, with: viewController)
+    }
```
Comment (hassaanelgarem) [inline WordPress/Classes/ViewRelated/Blog/Blog Dashboard/Cards/Quick Actions/DashboardQuickActionsCardCell.swift]: I don't think you need this. This cell no longer conforms to `BlogDashboardCardConfigurable` and `configureQuickActionButtons` get called directly in `BlogDashboardViewModel.dataSource`

File: WordPress/Classes/ViewRelated/Blog/Blog Dashboard/BlogDashboardViewController.swift [hunk 4]
```diff
@@ -147,7 +152,22 @@ final class BlogDashboardViewController: UIViewController {
...
+        NotificationCenter.default.addObserver(forName: .QuickStartTourElementChangedNotification, object: nil, queue: nil) { [weak self] notification in
+
+            if let info = notification.userInfo,
+               let element = info[QuickStartTourGuide.notificationElementKey] as? QuickStartTourElement {
+
+                switch element {
+                case .setupQuickStart, .removeQuickStart:
+                    self?.loadCardsFromCache()
+                case .stats:
+                    self?.mySiteScrollView?.scrollToTop(animated: true)
```
Comment (hassaanelgarem) [inline WordPress/Classes/ViewRelated/Blog/Blog Dashboard/BlogDashboardViewController.swift]: Re iPad scrolling issue: you can check here on `embeddedInScrollView` and either scroll `mySiteScrollView` or the `collectionView`

File: WordPress/Classes/ViewRelated/Blog/QuickStartTourGuide.swift [hunk 5]
```diff
@@ -154,7 +174,9 @@ open class QuickStartTourGuide: NSObject {
 
     private func addSiteMenuWayPointIfNeeded(for tour: QuickStartTour) -> QuickStartTour {
 
-        if currentTourOrigin == .blogDashboard && tour.shownInBlogDetails && !UIDevice.isPad() {
+        if currentEntryPoint == .blogDashboard &&
+            tour.possibleEntryPoints == [.blogDetails] &&
```
Comment (hassaanelgarem) [inline WordPress/Classes/ViewRelated/Blog/QuickStartTourGuide.swift]: [nit] What about adding a computed variable to `TourState` called `mustBeShownInBlogDetails` and then call it here and in `currentTourMustBeShownFromBlogDetails`?

File: WordPress/Classes/ViewRelated/Blog/QuickStartTourGuide.swift [hunk 6]
```diff
@@ -240,6 +263,14 @@ open class QuickStartTourGuide: NSObject {
         dismissCurrentNotice()
 
         guard let nextStep = getNextStep() else {
+
+            if let taskCompleteDescription = tourState.tour.taskCompleteDescription {
```
Comment (hassaanelgarem) [inline WordPress/Classes/ViewRelated/Blog/QuickStartTourGuide.swift]: Nice!

File: WordPress/Classes/ViewRelated/Blog/QuickStartTourGuide.swift [hunk 7]
```diff
@@ -285,19 +316,19 @@ open class QuickStartTourGuide: NSObject {
     static func customizeListTours(for blog: Blog) -> [QuickStartTour] {
         return [
             QuickStartCreateTour(),
-            QuickStartSiteTitleTour(blog: blog),
-            QuickStartSiteIconTour(),
-            QuickStartEditHomepageTour(),
+//            QuickStartSiteTitleTour(blog: blog),
+//            QuickStartSiteIconTour(),
+//            QuickStartEditHomepageTour(),
```
Comment (hassaanelgarem) [inline WordPress/Classes/ViewRelated/Blog/QuickStartTourGuide.swift]: Don't forget to revert these and the grow your audience ones 👍

File: WordPress/Classes/ViewRelated/Blog/QuickStartTours.swift [hunk 8]
```diff
@@ -31,9 +32,15 @@ extension QuickStartTour {
...
 
-    var shownInBlogDetails: Bool {
+    var taskCompleteDescription: NSAttributedString? {
         get {
-            return true
+            return nil
+        }
+    }
+
+    var possibleEntryPoints: Set<QuickStartTourEntryPoint> {
```
Comment (hassaanelgarem) [inline WordPress/Classes/ViewRelated/Blog/QuickStartTours.swift]: [?] I can see that almost all tours have the possible entry points specified. Also from my understanding having no possible entries and having both entries, ultimately get treated the same way.

I feel this ambiguity might cause human errors in the future when adding new tours.

I would suggest completely removing the default implementation. Or to keep it but return `[.blogDetails, .blogDashboard]`. I feel this is more verbose. Also, we can add a comment that we only need to override this if the tour can only be accessed from a specific place.

What do you think?

File: WordPress/Classes/ViewRelated/Blog/Blog Dashboard/BlogDashboardViewController.swift [hunk 9]
```diff
@@ -147,7 +152,22 @@ final class BlogDashboardViewController: UIViewController {
...
     private func addQuickStartObserver() {
-        NotificationCenter.default.addObserver(self, selector: #selector(loadCardsFromCache), name: .QuickStartTourElementChangedNotification, object: nil)
+        NotificationCenter.default.addObserver(forName: .QuickStartTourElementChangedNotification, object: nil, queue: nil) { [weak self] notification in
+
+            if let info = notification.userInfo,
+               let element = info[QuickStartTourGuide.notificationElementKey] as? QuickStartTourElement {
+
+                switch element {
+                case .setupQuickStart, .removeQuickStart:
+                    self?.loadCardsFromCache()
```
Comment (hassaanelgarem) [inline WordPress/Classes/ViewRelated/Blog/Blog Dashboard/BlogDashboardViewController.swift]: This removes tons of unnecessary reloading! Kudos on this!

**Review by hassaanelgarem — COMMENTED** [review]
Everything works as described except for a few minor issues:

On iPad when the view is scrollable (iPad air with all cards displayed), triggering the stats tour doesn't scroll to the top to show the quick action cells.

https://user-images.githubusercontent.com/25306722/165026005-19066e34-c2ca-4735-a025-063e0958ba6b.mp4

Another thing I noticed is that the "Task Complete" doesn't get dismissed automatically. It was mentioned in pbArwn-4dG-p2 that it should be auto-dismissable. Not sure if that's still the case. @osullivanchris thoughts?

**Comment by osullivanchris** [comment]
> Another thing I noticed is that the "Task Complete" doesn't get dismissed automatically. It was mentioned in pbArwn-4dG-p2 that it should be auto-dismissable. Not sure if that's still the case. @osullivanchris thoughts?

Yeah that's right! thanks @hassaanelgarem . Sorry if it was unclear.

I also think we should change the copy to say '...when you're ready for the next **task**'. I think we're using the word task a few times. But not the word tip. My bad.

File: WordPress/Classes/Extensions/UIScrollView+Helpers.swift (hunk 1 above)
Comment (momo-ozawa) [reply WordPress/Classes/Extensions/UIScrollView+Helpers.swift:84]: Fixed! Also, I ended up defining a `scrollHorizontallyToView` method -- which we can use for the Media task! 798360a

File: WordPress/Classes/ViewRelated/Blog/Blog Dashboard/BlogDashboardViewController.swift (hunk 2 above)
Comment (momo-ozawa) [reply WordPress/Classes/ViewRelated/Blog/Blog Dashboard/BlogDashboardViewController.swift]: Adding the `additionalSafeAreaInsets` ensures that the FAB is displayed above the notice 😄 

No additional insets | With additional insets
-- | --
 | 


However I noticed not all QS tasks do this. I think it would be good to have consistency across all the tasks. I can clean this up / tackle this in a separate PR -- I added this as a subtask

File: WordPress/Classes/ViewRelated/Blog/Blog Dashboard/Cards/Quick Actions/DashboardQuickActionsCardCell.swift (hunk 3 above)
Comment (momo-ozawa) [reply WordPress/Classes/ViewRelated/Blog/Blog Dashboard/Cards/Quick Actions/DashboardQuickActionsCardCell.swift]: Fixed! f198868 🙈 

Thanks for catching this! I have no idea how the configure ended back in the cell 🙃

File: WordPress/Classes/ViewRelated/Blog/Blog Dashboard/BlogDashboardViewController.swift (hunk 4 above)
Comment (momo-ozawa) [reply WordPress/Classes/ViewRelated/Blog/Blog Dashboard/BlogDashboardViewController.swift]: Fixed! 6fde18e

File: WordPress/Classes/ViewRelated/Blog/QuickStartTours.swift (hunk 8 above)
Comment (momo-ozawa) [reply WordPress/Classes/ViewRelated/Blog/QuickStartTours.swift]: Good point. I decided to completely remove the default implementation to make this property more intentional d4b9b41

File: WordPress/Classes/ViewRelated/Blog/QuickStartTourGuide.swift (hunk 5 above)
Comment (momo-ozawa) [reply WordPress/Classes/ViewRelated/Blog/QuickStartTourGuide.swift]: Good suggestion! Instead of adding it to `TourState,` I opted to add it to `QuickStartTour` and provide a default implementation 9f03e03

File: WordPress/Classes/ViewRelated/Blog/QuickStartTourGuide.swift (hunk 7 above)
Comment (momo-ozawa) [reply WordPress/Classes/ViewRelated/Blog/QuickStartTourGuide.swift]: 🙈 Fixed! b1e49c7

**Comment by momo-ozawa** [comment]
@osullivanchris @hassaanelgarem 

> Another thing I noticed is that the "Task Complete" doesn't get dismissed automatically. It was mentioned in pbArwn-4dG-p2 that it should be auto-dismissable.
>
> I also think we should change the copy to say '...when you're ready for the next task'

Fixed! a06c86d

Right now the auto-dismiss timer is set to 10 seconds (which matches the tour suggestion dismiss timer). Is this too long? Lmk what you think :)


https://user-images.githubusercontent.com/6711616/165084415-6ef386ed-fd60-4297-8987-ae7a9086af5a.mp4

**Comment by momo-ozawa** [comment]
@hassaanelgarem Thank you for a thorough review! Addressed your comments -- ready for a re-review! 🙇‍♀️

File: WordPress/Classes/Extensions/UIScrollView+Helpers.swift (hunk 1 above)
Comment (hassaanelgarem) [reply WordPress/Classes/Extensions/UIScrollView+Helpers.swift:84]: > I ended up defining a scrollHorizontallyToView method -- which we can use for the Media task

Thanks for that! It'll be super helpful!

File: WordPress/Classes/ViewRelated/Blog/Blog Dashboard/BlogDashboardViewController.swift (hunk 2 above)
Comment (hassaanelgarem) [reply WordPress/Classes/ViewRelated/Blog/Blog Dashboard/BlogDashboardViewController.swift]: Aaah okay got it!

File: WordPress/Classes/ViewRelated/Blog/Blog Dashboard/BlogDashboardViewController.swift (hunk 4 above)
Comment (hassaanelgarem) [reply WordPress/Classes/ViewRelated/Blog/Blog Dashboard/BlogDashboardViewController.swift]: I retested the iPad scrolling issue and I can verify it is indeed fixed 🚀

File: WordPress/Classes/ViewRelated/Blog/QuickStartTourGuide.swift [hunk 10]
```diff
@@ -285,6 +280,29 @@ open class QuickStartTourGuide: NSObject {
...
+            guard let self = self else {
+                return
+            }
+            ActionDispatcher.dispatch(NoticeAction.clearWithTag(self.noticeTag))
+            self.taskCompleteWorkItem?.cancel()
+            self.taskCompleteWorkItem = nil
+        }
+
+        DispatchQueue.main.asyncAfter(deadline: .now() + Constants.taskCompleteTimeout, execute: newWorkItem)
+        taskCompleteWorkItem = newWorkItem
```
Comment (hassaanelgarem) [inline WordPress/Classes/ViewRelated/Blog/QuickStartTourGuide.swift]: [nit] Instead of this, what about using `NoticeStyle.isDismissable`?

**Review by hassaanelgarem — COMMENTED** [review]
The task completion notice is now auto-dismissing as expected!

> Right now the auto-dismiss timer is set to 10 seconds (which matches the tour suggestion dismiss timer). Is this too long? 

I think it's a bit long, yes. We're waiting for the user's input for the tour suggestion, so it makes sense to make it long. But for the task completion notice, I feel it should be shorter. I suggest it be 5 seconds, same as other non-actionable notices.

**Comment by osullivanchris** [comment]
> I think it's a bit long, yes. We're waiting for the user's input for the tour suggestion, so it makes sense to make it long. But for the task completion notice, I feel it should be shorter. I suggest it be 5 seconds, same as other non-actionable notices.

Agree 5 seconds is fine for me. There's no critical information in this notice anyway. Its just an extra bonus piece of info if they want it.

File: WordPress/Classes/ViewRelated/Blog/QuickStartTourGuide.swift (hunk 10 above)
Comment (momo-ozawa) [reply WordPress/Classes/ViewRelated/Blog/QuickStartTourGuide.swift]: Done! 820a326

**Review by hassaanelgarem — APPROVED** [review]
:shipit:
